sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -114,13 +114,13 @@ class CacheManager extends Logging {
   }
 
   /**
-   * Un-cache all the cache entries that refer to the given plan.
+   * Un-cache the cache entry that refers to the given plan.
    */
   def uncacheQuery(spark: SparkSession, plan: LogicalPlan, blocking: Boolean): Unit = writeLock {
     val it = cachedData.iterator()
     while (it.hasNext) {
       val cd = it.next()
-      if (cd.plan.find(_.sameResult(plan)).isDefined) {
+      if (plan.sameResult(cd.plan)) {
```

Comment:
This is by design. This is for avoiding to get the incorrect results. See the original PR: https://github.com/apache/spark/pull/17097